File: AnYang/Array/35. Search Insert Position.py
# 2019-08-05 Mon

class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        
        l = len(nums)//2
        left = 0
        right = len(nums)-1
        
        while(left + 1 != right and left != right):
            if(target == nums[l]):
                return l
            if(target <= nums[l]):
                right = l
                l = (left + right)//2
            elif(target > nums[l]):
                left = l
                l = (left + right)//2
                
        if target > nums[right]:
            return right + 1
        if target <= nums[left]:
            return 0
        return right

File: tobehe/35.search-insert-position.cpp
/*
 * @lc app=leetcode id=35 lang=cpp
 *
 * [35] Search Insert Position
 */
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        
        int length = nums.size();

        int p0 = 0;
        int p1 = length - 1;

        while(p0 < p1)
        {
            int m = (p0 + p1 + 1)/2;
            if(target == nums[m])
                return m;
            else if(target < nums[m])
            {
                p1 = m - 1;
            }
            else
            {
                p0 = m + 1;
            }
        }   

        return target <= nums[p1]?p1:p1+1;
    }
};

